dhall-docs/src/Dhall/Docs/Store.hs

```diff
@@ -0,0 +1,42 @@
+{-| Utilities to interact with the dhall-docs home directory
+-}
+module Dhall.Docs.Store (getDocsHomeDirectory, makeHashForDirectory) where
+
+import Data.Map.Strict (Map)
+import Dhall.Crypto    (SHA256Digest (..), sha256Hash)
+import Path            (Abs, Dir, File, Path)
+import Path.IO         (XdgDirectory (..))
+
+import qualified Data.ByteString
+import qualified Data.Map.Strict as Map
+import qualified Path
+import qualified Path.IO
+
+{-| Fetches the dhall-docs home directory. If @XDG_DATA_HOME@ env var is
+    defined, then @${XDG_DATA_HOME}/dhall-docs@ will be returned. Otherwise,
+    "${HOME}/.local/share/dhall-docs"
+-}
+getDocsHomeDirectory :: IO (Path Abs Dir)
+getDocsHomeDirectory = do
+    dir <- Path.IO.getXdgDir Path.IO.XdgData $ Path.parseRelDir "dhall-docs"
+    Path.IO.ensureDir dir
+    return dir
+
+{-| Compute the hash for a directory. It takes into account the hierarchy
+    structure of it and the contents of its files, but not the name of the
+    actual files.
+
+    This is done by computing the hash of each file and sorting them by its
+    absolute file name, and computing the hash of the concatenation of all
+    hashes.
+-}
+makeHashForDirectory :: Path Abs Dir -> IO SHA256Digest
+makeHashForDirectory dir = do
+    -- Builds a map so key order is preserved between several calls
```

Comment:
I believe the hash does not change in some cases if the file name changes because the file names are discarded when computing the hash

However, there might be a simpler approach to do this, which is to create an archive from the directory (e.g. using the `tar` package, which supports doing it in memory) and then hash that.  That's essentially the same way that Nix hash directories (except that Nix has its own custom archive format).

Reply:
@Gabriel439 Thanks for your suggestion. I'm implementing it, but I see different hashes on different executions.

_I think_ its because the tarballs include the date of file creation of its files, which change on each execution obviously. A solution for this is to set that date to 0 (it is an epoch time). I tried it and now the hashes preserves, although the tar isn't an exact representation of the generated docs because of that change. Are you ok with that change? I'm pushing it so you can review it.

Reply:
Yeah, that's fine to set the date to 0.  This would actually be equivalent to what Nix does as well, too 🙂